design-proposals/klip-28-create-or-replace.md

```diff
@@ -0,0 +1,200 @@
+requirements change. To that end, all popular data stores have ways of managing and manipulating 
+existing data. For stream processing applications, a user may want to modify their application as 
+a result of:
+
+- Business Requirements: requirements simply change over time
+- Schema Evolution: the incoming data or required output has been modified
+- Optimizations: the same application can be executed more efficiently (either by user or engine)
+
+At time of writing, ksqlDB provides a crude mechanism for altering its application behavior: dropping 
+a query and restarting it at the earliest or latest offset. While this often works well in development 
+environments, there are limitations to its applicability in production:
+
+- Data Retention: the earliest available offset may not correspond with the beginning of time
+- Downtime: the delta between terminating and catch-up may be out of the application’s SLO
+- Output Routing: populating to the old output topic will produce duplicates while using a new output topic will require cascading migrations
+- Compute: recomputing the complete history for a query may not be feasible
+
+Kafka Streams provides more granular mechanisms (e.g. restarting queries with different behaviors 
+but identical consumer groups), but these methods burden users with extra complexity and lack guardrails.
+
+## Scope
+
+To better understand the scope of this KLIP and any future improvements, we define a taxonomy on 
+query upgrades as any combination of three types of characteristics: **source query, upgrade** and
+(optionally) **environment**:
+
+| **Category** | **Characteristic** | **Description** |
+|----------|----------------|-------------|
+| Query | Stateful | Stateful queries maintain local storage |
+| | Windowed | Windowed queries maintain a limited amount of state specified by a window in time
+| | Joined | Joined queries read from multiple sources
+| | Multistage | Multistage queries contain intermediate, non-user visible topics in Kafka
+| | Nondeterministic | Nondeterministic queries may produce different results when executing identical input
+| | Simple | Queries with none of the above characteristics
+| Upgrade | Transparent | Transparent upgrades change the way something is computer (e.g. improving a UDF performance)
+| | Data Selection | Data selecting query upgrades change which/how many events are emitted
+| | Schema Evolution | Schema evolving query upgrades change the output type of the data |
+| | Source Modifying | These upgrades change the source data, whether by means of modifying a JOIN or swapping out a source |
+| | Topology | These upgrades are invisible to the user, but change the topology, such as the number of sub-topologies or the ordering of operations (e.g. filter push down) |
+| | Scaling | Scaling upgrades change the physical properties of the query in order to enable better performance characteristics. |
```

Comment:
What's an example of this? Changing the number of streams threads, and other streams properties?

What about changing parameters such as the number of output partitions -- and other properties from the "WITH" clause, such as output format? Will that be supported?

Reply:
> What's an example of this?

I don't have a good example as available in the language today, but things like scaling out partitions or streams threads would such "scaling" operations. We could imagine changing number of stream threads might be one such thing (e.g. `CREATE OR REPLACE ... WITH(num_threads=1000)`)

The scope of what this KLIP will implement doesn't really address any scaling changes (such as changing number of output partitions) but that could be some advanced upgrades that we want to support in the future. I felt it was important to include this in the table though for completeness

Reply:
When a `CREATE OR REPLACE` statement is issued, does the new query use the query properties (e.g., number of streams threads) set at the time of the `CREATE OR REPLACE` statement or does it maintain properties from the original query (if present)? I assume the latter but think this is worth calling out explicitly. If this is the case, then changing the number of streams threads is already supported under this proposal.

Reply:
It would indeed use the most recent one - so I suppose it is already supported 😂 I can update the KLIP if you'd like, though perhaps just seeing this discussion is sufficient documentation

Reply:
Are we planning to do any validation on the new query properties, in terms of checking for compatibility? For example, I'm not entirely clear on what happens if a Streams app is restarted with a different value for `replication.factor` (assuming the internal topics are already created with the old value) but I suspect there are combinations of values where this might fail?

Reply:
> Are we planning to do any validation on the new query properties, in terms of checking for compatibility?

The intention for compatibility was to setup a whitelist of items that can change, and if anything else changes we consider the upgrade incompatible. Honestly, I hadn't considered the query properties until now, but they definitely seem something important to put into the compatibility calculus. 

As for your specific question, I think upping the replication factor is supported but reducing it is not (though I might be wrong on that one).

Reply:
Sounds good. Just wanted to check we had a plan :)

Reply:
I appreciate the thorough review 😄